File: go/internal/adapters/inventory.go
// Package adapters provides in-memory doubles for the application ports:
// the canonical design's InMemory/Fake implementations ARE the test doubles,
// so no mocking framework exists anywhere in this template.
package adapters

import (
	"context"
	"fmt"
	"sync"

	"warehouse/internal/application"
	"warehouse/internal/domain"
)

type heldUnits struct {
	sku    domain.Sku
	amount int64
}

// InMemoryInventoryGateway is a finite-stock InventoryGateway double backed
// by a per-SKU stock map. ReserveAll is atomic and the adapter is safe for
// concurrent use.
type InMemoryInventoryGateway struct {
	mu           sync.Mutex
	stock        map[domain.Sku]int64
	reservations map[string][]heldUnits
	FailRelease  bool
}

// NewInMemoryInventoryGateway starts from an optional initial stock map,
// copied defensively.
func NewInMemoryInventoryGateway(stock map[domain.Sku]int64) *InMemoryInventoryGateway {
	copied := make(map[domain.Sku]int64, len(stock))
	for sku, units := range stock {
		copied[sku] = units
	}
	return &InMemoryInventoryGateway{
		stock:        copied,
		reservations: make(map[string][]heldUnits),
	}
}

// SnapshotStock returns a copy of remaining units per SKU.
func (g *InMemoryInventoryGateway) SnapshotStock() map[domain.Sku]int64 {
	g.mu.Lock()
	defer g.mu.Unlock()
	out := make(map[domain.Sku]int64, len(g.stock))
	for sku, units := range g.stock {
		out[sku] = units
	}
	return out
}

// ReserveAll decrements stock for every line or none. A shortage wraps
// domain.InsufficientStock and leaves the map untouched. Identical
// idempotency keys replay the original reservation without a second debit.
func (g *InMemoryInventoryGateway) ReserveAll(
	_ context.Context,
	orderID domain.OrderID,
	lines []domain.OrderLine,
	idempotencyKey string,
) (application.ReservationToken, error) {
	g.mu.Lock()
	defer g.mu.Unlock()
	token := application.ReservationToken{OrderID: orderID, IdempotencyKey: idempotencyKey}
	if _, ok := g.reservations[idempotencyKey]; ok {
		return token, nil
	}
	needed := make([]heldUnits, 0, len(lines))
	for _, line := range lines {
		available := g.stock[line.SKU]
		if available < int64(line.Quantity) {
			return application.ReservationToken{}, fmt.Errorf(
				"reserve all: %w",
				domain.InsufficientStock{SKU: line.SKU, Requested: line.Quantity, Available: available},
			)
		}
		needed = append(needed, heldUnits{sku: line.SKU, amount: int64(line.Quantity)})
	}
	for _, held := range needed {
		g.stock[held.sku] -= held.amount
	}
	g.reservations[idempotencyKey] = needed
	return token, nil
}

// Release puts reserved units back. Unknown tokens are a no-op success.
func (g *InMemoryInventoryGateway) Release(_ context.Context, token application.ReservationToken) error {
	g.mu.Lock()
	defer g.mu.Unlock()
	if g.FailRelease {
		return fmt.Errorf(
			"release: %w",
			domain.CompensationFailure{Stage: "release", Detail: "forced failure"},
		)
	}
	held, ok := g.reservations[token.IdempotencyKey]
	if !ok {
		return nil
	}
	delete(g.reservations, token.IdempotencyKey)
	for _, item := range held {
		g.stock[item.sku] += item.amount
	}
	return nil
}

// compile-time proof that the adapter satisfies its port.
var _ application.InventoryGateway = (*InMemoryInventoryGateway)(nil)

File: swift/Sources/Warehouse/Adapters/Inventory.swift
import Dispatch

/// InventoryGateway double enforcing finite stock, for tests and demos.
public final class InMemoryInventoryGateway: InventoryGateway, @unchecked Sendable {
  private let queue = DispatchQueue(label: "warehouse.inventory")
  private var stock: [Sku: Int]
  private var reservations: [String: [(Sku, Int)]] = [:]
  public var failRelease = false

  public init(stock: [Sku: Int] = [:]) {
    self.stock = stock
  }

  public func snapshotStock() -> [Sku: Int] {
    queue.sync { stock }
  }

  public func reserveAll(
    orderId: OrderId,
    lines: [OrderLine],
    idempotencyKey: String
  ) -> Result<ReservationToken, InsufficientStock> {
    queue.sync {
      if reservations[idempotencyKey] != nil {
        return .success(ReservationToken(orderId: orderId, idempotencyKey: idempotencyKey))
      }
      var needed: [(Sku, Int)] = []
      for line in lines {
        let available = stock[line.sku] ?? 0
        if available < line.quantity.value {
          return .failure(
            InsufficientStock(
              sku: line.sku,
              requested: line.quantity,
              available: available
            )
          )
        }
        needed.append((line.sku, line.quantity.value))
      }
      for (sku, amount) in needed {
        stock[sku] = (stock[sku] ?? 0) - amount
      }
      reservations[idempotencyKey] = needed
      return .success(ReservationToken(orderId: orderId, idempotencyKey: idempotencyKey))
    }
  }

  public func release(token: ReservationToken) -> Result<Void, CompensationFailure> {
    queue.sync {
      if failRelease {
        return .failure(CompensationFailure(stage: "release", detail: "forced failure"))
      }
      guard let held = reservations.removeValue(forKey: token.idempotencyKey) else {
        return .success(())
      }
      for (sku, amount) in held {
        stock[sku] = (stock[sku] ?? 0) + amount
      }
      return .success(())
    }
  }
}
